### PR TITLE
chore: update lint tools versions + use ts 5

### PR DESCRIPTION
**Pull request by ccharly**
Updating `eslint` and other similar packages.

### CODE REVIEW

File: packages/keyring-eth-ledger-bridge/src/ledger-keyring.ts [hunk 1]
```diff
@@ -109,15 +115,15 @@ export class LedgerKeyring extends EventEmitter {
...
     return this.bridge.init();
   }
 
-  async destroy() {
+  async destroy(): Promise<void> {
     return this.bridge.destroy();
   }
 
-  async serialize() {
+  async serialize(): Promise<Partial<LedgerBridgeKeyringOptions>> {
```
Comment (ccharly) [inline packages/keyring-eth-ledger-bridge/src/ledger-keyring.ts:126]: I did this, but maybe we should have a dedicated type for the state here?

File: packages/keyring-eth-trezor/src/trezor-connect-bridge.ts [hunk 2]
```diff
@@ -35,29 +40,36 @@ export class TrezorConnectBridge implements TrezorBridge {
     this.trezorConnectInitiated = true;
   }
 
-  dispose() {
+  async dispose(): Promise<void> {
```
Comment (ccharly) [inline packages/keyring-eth-trezor/src/trezor-connect-bridge.ts:43]: Was not `async` but since it's returning a `Promise` it makes sense to make it `async`

File: packages/keyring-eth-trezor/src/trezor-keyring.ts [hunk 3]
```diff
@@ -116,14 +125,14 @@ export class TrezorKeyring extends EventEmitter {
     return this.bridge.model;
   }
 
-  init() {
+  async init(): Promise<void> {
```
Comment (ccharly) [inline packages/keyring-eth-trezor/src/trezor-keyring.ts:128]: Was not `async` but since it's returning a `Promise` it makes sense to make it `async`.

Also, the `KeyringController` is already `await`ing those calls: https://github.com/MetaMask/core/blob/main/packages/keyring-controller/src/KeyringController.ts#L2199-L2201

File: packages/keyring-eth-trezor/src/trezor-keyring.ts [hunk 4]
```diff
@@ -443,15 +453,16 @@ export class TrezorKeyring extends EventEmitter {
     });
   }
 
-  /**
-   * EIP-712 Sign Typed Data
-   */
+  // EIP-712 Sign Typed Data
```
Comment (ccharly) [inline packages/keyring-eth-trezor/src/trezor-keyring.ts:456]: Using a simple comment here, since most of those method are using this pattern

File: packages/keyring-snap-bridge/src/SnapKeyring.ts [hunk 5]
```diff
@@ -585,7 +589,7 @@ export class SnapKeyring extends EventEmitter {
    * @param snapId - The Snap ID to check allowed origins for.
    * @throws An error if the URL's origin is not in the Snap's allowed origins.
    */
-  #validateRedirectUrl(url: string, snapId: SnapId) {
+  async #validateRedirectUrl(url: string, snapId: SnapId): Promise<void> {
```
Comment (ccharly) [inline packages/keyring-snap-bridge/src/SnapKeyring.ts:592]: Seems non-necessary, but this is a private method, so making it `async` is ok here.

**Review by ccharly — COMMENTED** [review]
I did review (but cannot approve since I have some commits here). Left some comments with some open-questions + to explain some of the changes that were made by me.

File: jest.config.packages.js [hunk 6]
```diff
@@ -141,7 +141,7 @@ module.exports = {
   // runner: "jest-runner",
 
   // The paths to modules that run some code to configure or set up the testing environment before each test
-  setupFiles: ['../../tests/setup.ts'],
+  // setupFiles: ['../../tests/setup.ts'],
```
Comment (gantunesr) [inline jest.config.packages.js:144]: Why comment it out?

File: packages/keyring-api/src/rpc-handler.ts [hunk 7]
```diff
@@ -47,22 +47,22 @@ async function dispatchRequest(
   assert(request, JsonRpcRequestStruct);
 
   switch (request.method) {
-    case KeyringRpcMethod.ListAccounts: {
+    case `${KeyringRpcMethod.ListAccounts}`: {
```
Comment (gantunesr) [inline packages/keyring-api/src/rpc-handler.ts:50]: I'm not following on why this change is necessary, from a first look it seems redundant

File: packages/keyring-eth-ledger-bridge/src/ledger-hw-app.ts [hunk 8]
```diff
@@ -1,16 +1,8 @@
...
 import { Buffer } from 'buffer';
 
 import { GetAppNameAndVersionResponse } from './type';
 
-// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
-export interface MetaMaskLedgerHwAppEth extends LedgerHwAppEth {
-  openEthApp(): void;
-  closeApps(): void;
-  getAppNameAndVersion(): Promise<GetAppNameAndVersionResponse>;
-}
```
Comment (gantunesr) [inline packages/keyring-eth-ledger-bridge/src/ledger-hw-app.ts:12]: Just confirming that this is unused

File: packages/keyring-eth-ledger-bridge/src/ledger-keyring.ts [hunk 9]
```diff
@@ -36,6 +34,14 @@ type SignTransactionPayload = Awaited<
   ReturnType<LedgerBridge<LedgerIframeBridgeOptions>['deviceSignTransaction']>
 >;
 
+type AccountPageEntry = {
+  address: string;
+  balance: number | null;
+  index: number;
+};
+
+type AccountPage = AccountPageEntry[];
```
Comment (gantunesr) [inline packages/keyring-eth-ledger-bridge/src/ledger-keyring.ts:43]: Should these be exported?

File: packages/keyring-eth-ledger-bridge/test/document.shim.ts [hunk 10]
```diff
@@ -1,24 +1,31 @@
-// eslint-disable-next-line import/no-mutable-exports
+// eslint-disable-next-line import-x/no-mutable-exports
```
Comment (gantunesr) [inline packages/keyring-eth-ledger-bridge/test/document.shim.ts:1]: I'm not familiar with this one, how is it different from `import/no-mutable-exports`?

**Review by gantunesr — COMMENTED** [review]
Looking good, left a few comments

File: packages/keyring-eth-ledger-bridge/src/ledger-hw-app.ts (hunk 8 above)
Comment (ccharly) [reply packages/keyring-eth-ledger-bridge/src/ledger-hw-app.ts:12]: This is indeed unused and redundant here (while also be kinda invalid 🙈)

- The `interface` is using the same name than the `class` following it, meaning the `class` will override it anyway. I believe it has been introduced to force the methods signatures on the `class` itself
- `interface` can extends `class` to some extent but it seems to have some restrictions (specifically with private members)

File: packages/keyring-api/src/rpc-handler.ts (hunk 7 above)
Comment (ccharly) [reply packages/keyring-api/src/rpc-handler.ts:50]: `eslint was complaining here, most likely cause `request.method` is being infered as `string` (which makes sense given that `JsonRpcRequestStruct` is using `method: string()`.


For reference:
- https://typescript-eslint.io/rules/no-unsafe-enum-comparison/

I assume it's considered unsafe to compare a `string` with an `enum` value here, hence the use of template literal to go from (type-wise) `string === enum` to `string === string`

File: packages/keyring-eth-ledger-bridge/test/document.shim.ts (hunk 10 above)
Comment (ccharly) [reply packages/keyring-eth-ledger-bridge/test/document.shim.ts:1]: Looks like the plugin got renamed:

- https://github.com/un-ts/eslint-plugin-import-x/issues/24#issuecomment-1991605123

File: packages/keyring-eth-ledger-bridge/src/ledger-keyring.ts (hunk 9 above)
Comment (ccharly) [reply packages/keyring-eth-ledger-bridge/src/ledger-keyring.ts:43]: Neither `eslint` nor `tsc` complained about this type not being exported, but IMO we should since it's part of public functions signatures, lemme export them! Thanks!

File: jest.config.packages.js (hunk 6 above)
Comment (ccharly) [reply jest.config.packages.js:144]: We used the same pattern than for other configuration fields in this file:
- https://github.com/MetaMask/accounts/pull/55/files/54eb2a99f2b8bcf0396ac1cbe21621a8a70643fd#diff-67d7eaa7b8718c9f7f2fb10f5be717922c0175ecd68f4af4c3520e505580eb4bR146-R147

File: packages/keyring-eth-ledger-bridge/src/ledger-keyring.ts (hunk 9 above)
Comment (ccharly) [reply packages/keyring-eth-ledger-bridge/src/ledger-keyring.ts]: See:
- https://github.com/MetaMask/accounts/pull/55/commits/006878f9776c5ce4810e608175af8ab6883bc42b
- https://github.com/MetaMask/accounts/pull/55/commits/f916f75ce252c5336f995ad77e8d12e0eba56e77

File: packages/keyring-api/src/rpc-handler.ts (hunk 7 above)
Comment (danroc) [reply packages/keyring-api/src/rpc-handler.ts:50]: That's right, the type of the `request.method` field is `string` but the `case` values were `KeyringRpcMethod`, which caused the linter to complain.

I tought about changing the type of the `method` field, but then the linter would complain about the `default` case.

Another option I tried was to change this assertion:

```ts
assert(request, JsonRpcRequestStruct);
```

To something like (not the exact syntax):

```ts
assert(request, GetAccountRequestStruct | CreateAccountRequestStruct | ...);
```

But the type would not resolve correctly (I'm not sure if it's a Superstruct or TypeScript bug).

Comment (gantunesr) [reply packages/keyring-api/src/rpc-handler.ts:50]: Interesting, I would assume that the type of an `enum` member to be a specific type (`string`, `number`, `boolean`, etc). It does seems to be confusing but in a minor scale, it can be improved in a future PR.

File: packages/keyring-api/src/superstruct.test-d.ts [hunk 11]
```diff
@@ -4,6 +4,7 @@ import { expectAssignable, expectNotAssignable } from 'tsd';
 
 import { exactOptional, object } from '.';
 
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (gantunesr) [inline packages/keyring-api/src/superstruct.test-d.ts:7]: Should we keep this unused vars?

File: packages/keyring-api/src/superstruct.ts [hunk 12]
```diff
@@ -14,6 +14,7 @@ import type {
   Simplify,
 } from '@metamask/superstruct';
 
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (gantunesr) [inline packages/keyring-api/src/superstruct.ts:17]: Same here regarding unused vars

File: packages/keyring-eth-ledger-bridge/test/document.shim.ts [hunk 13]
```diff
@@ -1,24 +1,31 @@
...
 let documentShim: any;
 
+type Element = {
```
Comment (gantunesr) [inline packages/keyring-eth-ledger-bridge/test/document.shim.ts:4]: nit

```suggestion
type IframeElement = {
```

File: packages/keyring-eth-simple/src/simple-keyring.test.ts [hunk 14]
```diff
@@ -21,8 +21,10 @@ import {
 import { add0x, Hex } from '@metamask/utils';
 import assert from 'assert';
 import { keccak256 } from 'ethereum-cryptography/keccak';
+// eslint-disable-next-line @typescript-eslint/naming-convention
 import OldEthereumTx from 'ethereumjs-tx';
 
+// eslint-disable-next-line @typescript-eslint/naming-convention
 import SimpleKeyring from '.';
```
Comment (gantunesr) [inline packages/keyring-eth-simple/src/simple-keyring.test.ts:28]: Why is the import different here compared with the file `packages/keyring-eth-simple/src/sample.ts`?

```ts
import SimpleKeyring from './simple-keyring';
```

File: packages/keyring-eth-simple/src/simple-keyring.ts [hunk 15]
```diff
@@ -27,8 +27,14 @@ type KeyringOpt = {
   version?: SignTypedDataVersion | string;
 };
 
+type Wallet = {
+  privateKey: Buffer;
+  publicKey: Buffer;
+};
```
Comment (gantunesr) [inline packages/keyring-eth-simple/src/simple-keyring.ts:33]: nit suggestion,

```suggestion
type KeyPair = {
  privateKey: Buffer;
  publicKey: Buffer;
};
```

File: packages/keyring-eth-simple/src/simple-keyring.test.ts [hunk 16]
```diff
@@ -21,8 +21,10 @@ import {
 import { add0x, Hex } from '@metamask/utils';
 import assert from 'assert';
 import { keccak256 } from 'ethereum-cryptography/keccak';
+// eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (gantunesr) [inline packages/keyring-eth-simple/src/simple-keyring.test.ts:24]: What is the naming convention error here? Asking because I don't see a relevant difference with the imports from other files

**Review by gantunesr — COMMENTED** [review]
Minor nits!

File: packages/keyring-api/src/superstruct.test-d.ts (hunk 11 above)
Comment (ccharly) [reply packages/keyring-api/src/superstruct.test-d.ts:7]: We need it to infer the type `ExactOptionalObject` and this file is mostly used for type testing, so we need to keep it.

File: packages/keyring-api/src/superstruct.ts (hunk 12 above)
Comment (ccharly) [reply packages/keyring-api/src/superstruct.ts:17]: Same remark here, it's only used "statically" (for the `type ExactOptionalTag`, so we need to keep it

File: packages/keyring-eth-ledger-bridge/test/document.shim.ts (hunk 13 above)
Comment (ccharly) [reply packages/keyring-eth-ledger-bridge/test/document.shim.ts:4]: I must say I do prefer `Element` to match the `createElement` name here, is that ok for you?

File: packages/keyring-eth-simple/src/simple-keyring.ts (hunk 15 above)
Comment (ccharly) [reply packages/keyring-eth-simple/src/simple-keyring.ts:33]: Even though the name might be better, I think that we should keep the original name (this PR was meant to only fix eslint warning/errors, and this one is not). Is that ok?

File: packages/keyring-eth-simple/src/simple-keyring.test.ts (hunk 16 above)
Comment (ccharly) [reply packages/keyring-eth-simple/src/simple-keyring.test.ts:24]: I believe that's because "default" exports should be named using camelCase naming convention, but here it's using PascalCase instead.

You are allowed to use PascalCase if you use "destructuring import" (not sure what's the official name for this).

File: packages/keyring-eth-ledger-bridge/test/document.shim.ts (hunk 13 above)
Comment (gantunesr) [reply packages/keyring-eth-ledger-bridge/test/document.shim.ts:4]: is just a nit, not going to die on this hill

File: packages/keyring-eth-simple/src/simple-keyring.ts (hunk 15 above)
Comment (gantunesr) [reply packages/keyring-eth-simple/src/simple-keyring.ts:33]: sure

File: packages/keyring-eth-simple/src/simple-keyring.test.ts (hunk 14 above)
Comment (ccharly) [reply packages/keyring-eth-simple/src/simple-keyring.test.ts:28]: I did change the one in `sample.ts` to match the original import path!

**Review by gantunesr — APPROVED** [review]
